guppylang/checker/expr_checker.py

```diff
@@ -188,13 +192,36 @@ def _synthesize(
         return ExprSynthesizer(self.ctx).synthesize(node, allow_free_vars)
 
     def visit_Tuple(self, node: ast.Tuple, ty: Type) -> tuple[ast.expr, Subst]:
-        if not isinstance(ty, TupleType) or len(ty.element_types) != len(node.elts):
+        if not (isinstance(ty, TupleType) and len(ty.element_types) == len(node.elts)):
             return self._fail(ty, node)
-        subst: Subst = {}
-        for i, el in enumerate(node.elts):
-            node.elts[i], s = self.check(el, ty.element_types[i].substitute(subst))
-            subst |= s
-        return node, subst
+
+        # Tuples can either be inert python tuples or tuples of functions which
+        # can be called in guppy. The former thing is checkable, but in the
+        # latter case we should be able to synthesise function types for the
+        # elements. Check here whether the given type is a tuple of function
+        # types to work out which case we're in.
```

Comment:
Is this distinction needed anymore? Can we just revert the code to how it was before?

Reply:
It seems this is needed to make the row semantics of tuples work, i.e. having `(T) == T` for tuples of functions, but I think it's simpler to delete this and accept this behaviour

Reply:
Agreed!